src/main/java/io/reactivex/processors/UnicastProcessor.java

```diff
@@ -38,380 +41,323 @@
-        volatile boolean done;
-        /** 
-         * The terminal error if not null. 
-         * Must be set before writing to done and read after done == true.
-         */
-        Throwable error;
+        for (;;) {
 
-        /** Set to 1 atomically for the first and only Subscriber. */
-        final AtomicBoolean once = new AtomicBoolean();
-        
-        /** 
-         * Called when the Subscriber has called cancel.
-         * This allows early termination for those who emit into this
-         * subject so that they can stop immediately 
-         */
-        Runnable onCancelled;
-        
-        /**
-         * Constructs the state with the given capacity and optional cancellation callback.
-         * @param capacityHint the capacity hint for the internal buffer
-         * @param onCancelled the optional cancellation callback
-         */
-        public State(int capacityHint, Runnable onCancelled) {
-            this.onCancelled = onCancelled;
-            queue = new SpscLinkedArrayQueue<T>(capacityHint);
-        }
-        
-        @Override
-        public void subscribe(Subscriber<? super T> s) {
-            if (!once.get() && once.compareAndSet(false, true)) {
-                s.onSubscribe(this);
-                subscriber.lazySet(s); // full barrier in drain
-                if (cancelled) {
-                    subscriber.lazySet(null);
+            long r = requested.get();
+            long e = 0L;
+            
+            while (r != e) {
+                boolean d = done;
+                
+                T t = q.poll();
+                boolean empty = t == null;
+                
+                if (checkTerminated(d, empty, a, q)) {
                     return;
                 }
-                drain();
-            } else {
-                if (done) {
-                    Throwable e = error;
-                    if (e != null) {
-                        EmptySubscription.error(e, s);
-                    } else {
-                        EmptySubscription.complete(s);
-                    }
-                } else {
-                    EmptySubscription.error(new IllegalStateException("Only a single subscriber allowed."), s);
+                
+                if (empty) {
+                    break;
                 }
-            }
-        }
-        
-        @Override
-        public void request(long n) {
-            if (SubscriptionHelper.validateRequest(n)) {
-                BackpressureHelper.add(requested, n);
-                drain();
-            }
-        }
-        
-        @Override
-        public void cancel() {
-            if (!cancelled) {
-                cancelled = true;
                 
-                if (getAndIncrement() == 0) {
-                    clear(queue);
+                a.onNext(t);
+                
+                e++;
+            }
+            
+            if (r == e) {
+                if (checkTerminated(done, q.isEmpty(), a, q)) {
```

Comment:
There's also a [PMD check](http://pmd.github.io/pmd-4.2.5/rules/basic.html#CollapsibleIfStatements) for this one